gdnative-core/src/init/property/hint.rs

```diff
@@ -0,0 +1,366 @@
+//! Strongly typed property hints.
+
+use std::fmt;
+use std::ops::RangeInclusive;
+
+use crate::GodotString;
+use crate::VariantType;
+
+use super::ExportInfo;
+
+/// Hints that an integer or float property should be within an inclusive range.
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```rust
+/// use gdnative_core::init::property::hint::RangeHint;
+///
+/// let hint: RangeHint<f64> = RangeHint::new(0.0, 20.0).or_greater();
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Default)]
+pub struct RangeHint<T> {
+    /// Minimal value, inclusive
+    pub min: T,
+    /// Maximal value, inclusive
+    pub max: T,
+    /// Optional step value for the slider
+    pub step: Option<T>,
+    /// Allow manual input above the `max` value
+    pub or_greater: bool,
+    /// Allow manual input below the `min` value
+    pub or_lesser: bool,
+}
+
+impl<T> RangeHint<T> {
+    /// Creates a new `RangeHint`.
+    pub fn new(min: T, max: T) -> Self {
+        RangeHint {
+            min,
+            max,
+            step: None,
+            or_greater: false,
+            or_lesser: false,
+        }
+    }
+
+    /// Builder-style method that returns `self` with the given step.
+    pub fn with_step(mut self, step: T) -> Self {
+        self.step.replace(step);
+        self
+    }
+
+    /// Builder-style method that returns `self` with the `or_greater` hint.
+    pub fn or_greater(mut self) -> Self {
+        self.or_greater = true;
+        self
+    }
+
+    ///  Builder-style method that returns `self` with the `or_lesser` hint.
+    pub fn or_lesser(mut self) -> Self {
+        self.or_lesser = true;
+        self
+    }
+}
+
+impl<T> From<RangeInclusive<T>> for RangeHint<T> {
+    fn from(range: RangeInclusive<T>) -> Self {
+        let (min, max) = range.into_inner();
+        RangeHint::new(min, max)
+    }
+}
+
+impl<T: fmt::Display> fmt::Display for RangeHint<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{},{}", self.min, self.max)?;
+        if let Some(step) = &self.step {
+            write!(f, ",{}", step)?;
+        }
+
+        if self.or_greater {
+            write!(f, ",or_greater")?;
+        }
+        if self.or_lesser {
+            write!(f, ",or_lesser")?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Hints that an integer, float or string property is an enumerated value to pick in a list.
+///
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```rust
+/// use gdnative_core::init::property::hint::EnumHint;
+///
+/// let hint = EnumHint::new(&["Foo", "Bar", "Baz"]);
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Default)]
+pub struct EnumHint {
+    values: &'static [&'static str],
```

Comment:
Could this be made into a `Vec<String>`? Having it this way means it is way harder to have any kind of dynamic/runtime loading of enum variants.

I don't know how much of a difference it makes in practice and how common such a use-case could be, but since export information is usually only provided once this should not have any significant performance impact.